### PR TITLE
Balder/monitor 2 mutex and cond 4

### PR DESCRIPTION
**Pull request by baldersheim**
@vekterli PR

### CODE REVIEW

**Review by vekterli — APPROVED** [review]
:+1: see comment about timeout semantics. Extra :+1: for moving more code to monotonic clocks!

File: document/src/vespa/document/util/queue.h [hunk 1]
```diff
@@ -16,35 +16,32 @@ class Semaphore
...
-            vespalib::MonitorGuard guard(_sync);
-            if (_numWaiters == 0) break;
-            _count++;
-            guard.signal();
+            {
+                std::lock_guard guard(_lock);
+                if (_numWaiters == 0) break;
+                _count++;
+            }
+            _cond.notify_one();
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/util/queue.h:32]: Agree, I think an assert is better.
That will make sure we fail fast if we have logical bugs.

File: messagebus/src/vespa/messagebus/routablequeue.cpp [hunk 2]
```diff
@@ -43,9 +44,9 @@ RoutableQueue::dequeue(duration timeout)
 {
     steady_clock::time_point startTime = steady_clock::now();
     duration left = timeout;
-    vespalib::MonitorGuard guard(_monitor);
+    std::unique_lock guard(_lock);
     while (_queue.size() == 0 && left > duration::zero()) {
-        if (!guard.wait(left) || _queue.size() > 0) {
+        if ((_cond.wait_for(guard, left) == std::cv_status::no_timeout) || (_queue.size() > 0)) {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/routablequeue.cpp]: Yes, good catch.

File: messagebus/src/vespa/messagebus/messenger.cpp [hunk 3]
```diff
@@ -157,22 +156,23 @@ class DiscardRecurrentTasks : public SyncTask {
...
-    _closed(false),
-    _skip_request_thread(skip_request_thread),
-    _skip_reply_thread(skip_reply_thread)
+        _lock(),
+        _pool(128000),
+        _children(),
+        _queue(),
+        _closed(false),
+        _skip_request_thread(skip_request_thread),
+        _skip_reply_thread(skip_reply_thread)
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/messenger.cpp]: Oops, I guess clion played me a little trick...

**Pull request merged**